sources/core/Xenko.Core/Threading/ConcurrentFixedPool.cs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) Xenko contributors (https://xenko.com)
+// Distributed under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System;
+using System.Threading;
+
+namespace Xenko.Core.Threading
+{
+    /// <summary>
+    /// A stack of fixed size, that size must be a power of two,
+    /// always contains at least one item.
+    /// Fastest collection type under heavy concurrency.
+    /// </summary>
+    public class ConcurrentFixedPool<T> where T : class, new()
```

Comment:
Wondering if `new()` generic constraint should be removed. Its only purpose here is to be able to do `new T()` on lines 37 and 55.

As far as I can remember, that constraint is translated by the compiler with using `Activator.CreateInstance` which uses reflection and is slower. The following article seems to confirm that: https://devblogs.microsoft.com/premier-developer/dissecting-the-new-constraint-in-c-a-perfect-example-of-a-leaky-abstraction/

@Eideren can you try to remove it and instead pass a *factory* method to the ConcurrentFixedPool constructor such as:

```csharp
// constructor signature
public ConcurrentFixedPool(int size, Func<T> factory) { }

// call
var pool = new ConcurrentFixedPool<T>(42, () => new T());
```

Then check whether performance and/or allocations are changed between the two versions.

Reply:
It's actually far slower, I tested it before and that's why I defaulted to new() instead

Reply:
Weird. The article from the Microsoft engineer seems to indicate the opposite.

Reply:
I'll try it again just in case I missed something.

Reply:
I'm pretty sure I didn't test it under release at the time, it's hard to loose the habit of just F5'ing after any changes. Anyway, in debug mode the average runtime of a pop->push is about 2% slower with a factory. The run to run variant is not even half a percent so that 2% is not within margin of error.
In release, factory is about 6% faster. I don't test instance creation specifically because it is supposed to be an exceptional case, this type is supposed to always have more items than required so that branch is not supposed to be hit, it is only there as a safeguard to keep the logic working. I'll push those changes asap.